pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentUploadDownloadRestletResource.java

```diff
@@ -403,6 +421,156 @@ private SuccessResponse uploadSegment(@Nullable String tableName, TableType tabl
+        String sourceDownloadURIStr = segmentMetadataInfo.getSegmentDownloadURI();
+        if (StringUtils.isEmpty(sourceDownloadURIStr)) {
+          throw new ControllerApplicationException(LOGGER,
+              "'DOWNLOAD_URI' is required as a field within the multipart object for METADATA batch upload mode.",
+              Response.Status.BAD_REQUEST);
+        }
+        // The downloadUri for putting into segment zk metadata
+        String segmentDownloadURIStr = sourceDownloadURIStr;
+
+        String tempFileName = TMP_DIR_PREFIX + UUID.randomUUID();
+        tempEncryptedFile = new File(provider.getFileUploadTempDir(), tempFileName + ENCRYPTED_SUFFIX);
+        tempFiles.add(tempEncryptedFile);
+        tempDecryptedFile = new File(provider.getFileUploadTempDir(), tempFileName);
+        tempFiles.add(tempDecryptedFile);
+        tempSegmentDir = new File(provider.getUntarredFileTempDir(), tempFileName);
+        tempFiles.add(tempSegmentDir);
+
+        boolean encryptSegment = StringUtils.isNotEmpty(crypterClassNameInHeader);
+        File destFile = encryptSegment ? tempEncryptedFile : tempDecryptedFile;
+        // override copySegmentToFinalLocation if override provided in headers:COPY_SEGMENT_TO_DEEP_STORE
+        // else set to false for backward compatibility
+        String copySegmentToDeepStore = extractHttpHeader(headers,
+            FileUploadDownloadClient.CustomHeaders.COPY_SEGMENT_TO_DEEP_STORE);
+        boolean copySegmentToFinalLocation = Boolean.parseBoolean(copySegmentToDeepStore);
+        createSegmentFileFromSegmentMetadataInfo(segmentMetadataInfo, destFile);
+        if (encryptSegment) {
+          decryptFile(crypterClassNameInHeader, tempEncryptedFile, tempDecryptedFile);
+        }
+
+        String metadataProviderClass = DefaultMetadataExtractor.class.getName();
+        SegmentMetadata segmentMetadata = getSegmentMetadata(tempDecryptedFile, tempSegmentDir, metadataProviderClass);
+        LOGGER.info("Processing upload request for segment: {} of table: {} with upload type: {} from client: {}, "
+                + "ingestion descriptor: {}", segmentName, tableNameWithType, uploadType, clientAddress,
+            ingestionDescriptor);
+
+        // Validate segment
+        if (tableConfig.getIngestionConfig() == null || tableConfig.getIngestionConfig().isSegmentTimeValueCheck()) {
+          SegmentValidationUtils.validateTimeInterval(segmentMetadata, tableConfig);
+        }
+        // TODO: Include the un-tarred segment size when using the METADATA push rest API. Currently we can only use the
+        //  tarred segment size as an approximation.
+        long segmentSizeInBytes = getSegmentSizeFromFile(sourceDownloadURIStr);
+        long untarredSegmentSizeInBytes = 0L;
+        if (segmentSizeInBytes > 0) {
+          untarredSegmentSizeInBytes = segmentSizeInBytes;
+        }
+        SegmentValidationUtils.checkStorageQuota(segmentName, untarredSegmentSizeInBytes, tableConfig, _controllerConf,
```

Comment:
This only checks the quota I guess, but when we have a batch 100 segments, quota does not get updated as we check for every segment in this loop? A segment actually has to be added to the table for this to reflect in quota. In this case we will let in a batch of segments. 

cc @shounakmk219

Reply:
This observation is correct and quota check is not evaluated correctly in batch mode. Couple of issues here:
1) quota check can be moved inside the if block prior to calling `zkOperator.completeSegmentsOperations`
2) The segment size used here is the metadata file size as opposed to the actual segment size. This is the case for the existing single segment upload as well as the batch mode.